src/boot/efi/pe.c

```diff
@@ -160,6 +166,52 @@ static void locate_sections(
         }
 }
 
+static UINT32 get_compatibility_entry_address(const struct DosFileHeader *dos, const struct PeFileHeader *pe) {
+        UINTN addr = 0, size = 0;
+        static const CHAR8 *sections[] = { (CHAR8 *) ".compat", NULL };
+
+        /* The kernel may provide alternative PE entry points for different PE architectures. This allows
+         * booting a 64bit kernel on 32bit EFI that is otherwise running on a 64bit CPU. The locations of any
+         * such compat entry points are located in a special PE section. */
+
+        locate_sections((const struct PeSectionHeader *) ((const UINT8 *) dos + section_table_offset(dos, pe)),
+                        pe->FileHeader.NumberOfSections,
+                        sections,
+                        &addr,
+                        NULL,
+                        &size);
+
+        if (size == 0)
+                return 0;
+
+        typedef struct {
+                UINT8 type;
+                UINT8 size;
+                UINT16 machine_type;
+                UINT32 entry_point;
+        } _packed_ LinuxPeCompat1;
+
+        while (size >= sizeof(LinuxPeCompat1)) {
+                LinuxPeCompat1 *compat = (LinuxPeCompat1 *) ((UINT8 *) dos + addr);
```

Comment:
according to C specs you aren't even allowed to to *take* an unaligned pointer of a type, or you are already in undefined behaviour territory. (i.e. the compiler may assume that a pointer of type LinuxPeCompat* is always aligned, and hence the check below is something it can optimize away, since it already knows it is, it after all has the right type).



i.e. you need to check for alignment *before* you cast to the struct, not after.